lib/translator/translation.cc

```diff
@@ -69,39 +70,73 @@ BeginResponse Translation::Begin(Span<const uint8_t> scsi_cmd,
+CompleteResponse Translation::Complete(
+    Span<const nvme::GenericQueueEntryCpl> cpl_data, Span<uint8_t> buffer_in,
+    Span<uint8_t> sense_buffer) {
+  CompleteResponse resp = {};
   if (pipeline_status_ == StatusCode::kUninitialized) {
     DebugLog("Invalid use of API: Complete called before Begin");
-    return ApiStatus::kFailure;
+    resp.status = ApiStatus::kFailure;
+    return resp;
   }
-  if (pipeline_status_ == StatusCode::kFailure) {
-    // TODO fill buffer with SCSI CHECK CONDITION response
+
+  if (cpl_data.size() != nvme_cmd_count_) {
+    DebugLog(
+        "Invalid use of API, completion count %u does not equal command count "
+        "%u",
+        cpl_data.size(), nvme_cmd_count_);
+    resp.status = ApiStatus::kFailure;
+    return resp;
+  }
+
+  if (pipeline_status_ != StatusCode::kSuccess) {
+    ScsiStatus scsi_status = {
+        .status = scsi::Status::kCheckCondition,
+        .sense_key = scsi::SenseKey::kIllegalRequest,
+        .asc = scsi::AdditionalSenseCode::kInvalidFieldInCdb,
+        .ascq = scsi::AdditionalSenseCodeQualifier::kNoAdditionalSenseInfo};
+    FillSenseBuffer(sense_buffer, scsi_status);
     AbortPipeline();
-    return ApiStatus::kSuccess;
+    resp.status = ApiStatus::kSuccess;
+    resp.scsi_status = scsi_status.status;
+    return resp;
+  }
+
+  // Verify NVMe commands completed successfully. If not translate error.
+  for (uint32_t i = 0; i < cpl_data.size(); ++i) {
+    const nvme::GenericQueueEntryCpl& cpl_entry = cpl_data[i];
+    const nvme::CplStatus& cpl_status = cpl_entry.cpl_status;
+    ScsiStatus scsi_status = StatusToScsi(cpl_status.sct, cpl_status.sc);
+    if (scsi_status.status != scsi::Status::kGood) {
+      FillSenseBuffer(sense_buffer, scsi_status);
+      AbortPipeline();
+      resp.status = ApiStatus::kSuccess;
+      resp.scsi_status = scsi_status.status;
+      return resp;
+    }
   }
 
   // Switch cases should not return
-  ApiStatus ret;
+  resp.status = ApiStatus::kSuccess;
   Span<const uint8_t> scsi_cmd_no_op = scsi_cmd_.subspan(1);
   scsi::OpCode opc = static_cast<scsi::OpCode>(scsi_cmd_[0]);
-
   switch (opc) {
     case scsi::OpCode::kInquiry:
-      pipeline_status_ = InquiryToScsi(scsi_cmd_no_op, buffer, GetNvmeCmds());
-      ret = ApiStatus::kSuccess;
+      pipeline_status_ =
+          InquiryToScsi(scsi_cmd_no_op, buffer_in, GetNvmeCmds());
       break;
     case scsi::OpCode::kReadCapacity10:
-      pipeline_status_ = ReadCapacity10ToScsi(buffer, nvme_cmds_[0]);
-      ret = ApiStatus::kSuccess;
+      pipeline_status_ = ReadCapacity10ToScsi(buffer_in, nvme_cmds_[0]);
       break;
     case scsi::OpCode::kRequestSense:
-      pipeline_status_ = RequestSenseToScsi(scsi_cmd_no_op, buffer);
-      ret = ApiStatus::kSuccess;
+      pipeline_status_ = RequestSenseToScsi(scsi_cmd_no_op, buffer_in);
+      break;
+    default:
+      DebugLog("Invalid case reached. Please contact SCSI2NVMe devs.");
```

Comment:
Would it make sense to log the value of the unexpected opcode?

Reply:
Good idea, I think that would be useful if this error ever occurs.